### DISCUSSION TITLE
How to track type coverage

### DISCUSSION THREAD

**Discussion by alies-dev** [vimeo/psalm#8127] [Q&A] [answered]
I have an idea of tracking type coverage on private project that uses GitHub and GitHub actions. I've decided to start from a simple setup to store info and build charts on google spreadsheet but the a problem to extract type coverage on GitHub workflow. Does anyone solved similar problem or used another approach to track type coverage?

example:
<img width="1627" alt="image" src="https://user-images.githubusercontent.com/5278175/174592564-97f46f01-0d50-4ca6-898a-0054a4c93bf3.png">

**Comment by orklah** [comment, marked as answer]
Maybe take a look at [shepherd](https://github.com/psalm/shepherd). It's what Psalm use to contact shepherd.dev and store and display widgets you see on: https://shepherd.dev/

The request to shepherd is built in Psalm, but it's possible to override the host name to target any domain.

So, you could receive on your website a payload with the summary of Psalm analysis after each CI run (either you host the whole shepherd repo on it or you just create the same entry point only to receive the datas)

See here how the payload is constructed:
https://github.com/vimeo/psalm/blob/3e837cabd047e3508543378d291094826f0772e8/src/Psalm/Plugin/Shepherd.php#L37

and here to override the host name:
https://github.com/vimeo/psalm/blob/fccc366c1eb49dcab59b662d2bd7f20f7a0b692a/src/Psalm/Internal/Cli/Psalm.php#L300

(shepherd is actually an ordinary psalm plugin built in Psalm itself. If you're not satisfied with the format or the way it's done, you could perfectly create your own plugin base on the Shepherd class to change it however you like)

**Reply by alies-dev** [reply]
thanks for the great tip, @orklah , it works! I just specified `--shepherd=https://script.google.com/macros/s/XXX/exec` param and updated request parser on google app level 👍 . Unfortunately this report doesn't contain info about a number of issues, probably I'll create a new plugin for these purposes. Or do you think it's beneficial to add this info to shepherd plugin? - In this case I can create a PR.

A note for other devs: , it doesn't work locally: shepherd sends this data only if there are some env variables available on CI, see `\Psalm\Internal\ExecutionEnvironment\BuildInfoCollector::collect` for more details.

PS: I'll probably write an article about my case when I'll polish my solution. And then I'll post it here.

**Reply by AndrolGenhald** [reply]
If you want to do a PR adding documentation to walk through this that would be excellent!

**Reply by orklah** [reply]
Great to see it works fine :)

I have no problem with adding new params in shepherd payload itself as long as it doesn't break compatibility with shepherd.dev. number of issues, time and memory seems like great additions we could even display one day on the website (though I'm not forcing you to change shepherd if you don't want, please just create an issue in the repo so we know it's on the todo list)

**Reply by alies-dev** [reply]
@orklah 
one more question: `/hooks/psalm` query [is hardcoded](https://github.com/vimeo/psalm/blob/97e6511fab6622c1ca434c665dd895e77a0eeca2/src/Psalm/Plugin/Shepherd.php#L88) on Shepherd what makes this plugin not fully flexible: you can specify a custom domain only, not a full URL. What do you think about breaking backward compatibility in Psalm v5 and using full URL for `--shepherd` option? It will allow using domains and services where you can't change URLs, in my case this is google apps + google spreadsheets combination.

**Reply by orklah** [reply]
I'd be fine with a PR for that :)

**Reply by alies-dev** [reply]
>I'd be fine with a PR for that :)

PR https://github.com/vimeo/psalm/pull/9133
